### PR TITLE
Silence docs

### PR DESCRIPTION
**Pull request by therealsilence**
- Significant revisions to "Storm Basics" for additional background
  - added operator chaining / "order matters"
  - added syntax conventions for use throughout reference docs
  - expanded / clarified differences between operator and macro syntax

- Added reference guide for Storm data modification operators
- Added framework for additional Storm reference guides

### CODE REVIEW

File: docs/synapse/userguides/ug014_storm_ref_filter.rst [hunk 1]
```diff
@@ -0,0 +1,20 @@
+Storm Reference - Fitler Operators and Comparators
```
Comment (vEpiphyte) [inline docs/synapse/userguides/ug014_storm_ref_filter.rst]: spelling

File: docs/synapse/userguides/ug011_storm_basics.rst [hunk 2]
```diff
@@ -93,13 +64,158 @@ Working with Synapse data commonly involves three broad types of operations:
...
+| DNS A records to  |                                    |                                |   for "pivot" operator          |
+| the domains se    |                                    |                                |                                 |
+| associated with   |                                    |                                |                                 |
+| those records     |                                    |                                |                                 |
++-------------------+------------------------------------+--------------------------------+---------------------------------+ 
+
+**Note:** Filtering (including or excluding a subset of results) is a common operation in Storm, supported by the ``filter()`` operator. Storm supports an extensive and flexible set of filtering criteria, and as a result the full operator syntax for filtering is generally complex and therefore impractical to use at the CLI. For this reason, filter operations at the CLI are generally performed using the macro syntax, as shown above.
```
Comment (vEpiphyte) [inline docs/synapse/userguides/ug011_storm_basics.rst]: Per discussion - filter() isn't a usable command as far as I can tell / test

File: docs/synapse/userguides/ug011_storm_basics.rst [hunk 3]
```diff
@@ -93,13 +64,158 @@ Working with Synapse data commonly involves three broad types of operations:
...
+
+See the Storm reference guides <link> for a detailed discussion of individual operators and their operator (full) vs. macro syntax.
```
Comment (vEpiphyte) [inline docs/synapse/userguides/ug011_storm_basics.rst]: missing link

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+The operators below can be used to modify the Synapse hypergraph by:
+
+* adding or deleting nodes
+* adding cross-reference (xref) nodes
+* setting or modifying properties on nodes
+* adding or deleting tags from nodes
+
+(**Note:** currently there is no operator for deleting properties from nodes (e.g., ``delprop()``). This is planned for a future release.)
+
+All of these operators are defined in `storm.py`__.
```
Comment (vEpiphyte) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: link is broken

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+
+**IMPORTANT:** Synapse does not have an "are you sure?" prompt. Caution should be used with operators that can modify Synapse data, especially when used on the output of complex queries that may modify (or delete) large numbers of nodes. It is **strongly recommended** that you validate the output of a query by first running the query on its own to ensure it returns the expected results before applying any operator that will modify that data.
+
+See the `Storm Syntax Conventions`__ for an explanation of the usage format used below.
```
Comment (vEpiphyte) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: link is broken - should link to .html and not .rst

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+* adding or deleting nodes
+* adding cross-reference (xref) nodes
+* setting or modifying properties on nodes
+* adding or deleting tags from nodes
+
+(**Note:** currently there is no operator for deleting properties from nodes (e.g., ``delprop()``). This is planned for a future release.)
+
+All of these operators are defined in `storm.py`__.
+
+**IMPORTANT:** Synapse does not have an "are you sure?" prompt. Caution should be used with operators that can modify Synapse data, especially when used on the output of complex queries that may modify (or delete) large numbers of nodes. It is **strongly recommended** that you validate the output of a query by first running the query on its own to ensure it returns the expected results before applying any operator that will modify that data.
```
Comment (vEpiphyte) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: This is better as an admonition - probably a 'danger' or 'warning' directive.
http://docutils.sourceforge.net/docs/ref/rst/directives.html#admonitions

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+
+*Node with Properties:*
+
+``addnode( inet:dns:a , ( woot.com , 1.2.3.4 ) , :seen:min = "2017-08-01 01:23" , :seen:max = "2017-08-10 04:56" )``
+
+``[ inet:dns:a = ( woot.com , 1.2.3.4 ) :seen:min = "2017-08-01 01:23" :seen:max = "2017-08-10 04:56" ]``
+
+**Usage Notes:**
+
+* ``addnode()`` used at the Synapse CLI is most suitable for adding a relatively small number of nodes. For larger amounts of data, it is preferable to use the Synapse `ingest`__ subsystem to automate the process.
```
Comment (vEpiphyte) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: link is broken - should link to .html and not .rst

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+**Macro Syntax Notes:**
+
+* The macro syntax can create as many nodes as are specified within the brackets.
+* All nodes specified within the brackets that do not already exist will be created. For nodes that already exist, Synapse will return data for that node.
+
+
+addxref()
+---------
+
+Create one or more cross-reference (`xref`__) nodes that reference a specified ``<form>=<valu>``.
```
Comment (vEpiphyte) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: link is broken - should link to .html and not .rst

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+
+delnode()
+---------
+
+Deletes the specified node(s) from a Cortex.
+
+**Operator Syntax:**
+
+.. parsed-literal::
+  *<query>* **delnode(force=1)**
```
Comment (vEpiphyte) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: force is not required in order to run delnode.  force=1 is an optional arg that actually causes the node deletion to occur.  Otherwise this operator just consumes the current nodes in the query and does nothing with them.

File: docs/synapse/userguides/ug013_storm_ref_lift.rst [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+guid()
+------
+Todo
+
+alltag()
+--------
+Todo
+
+.. _storm.py: ../../../synapse/lib/storm.py
```
Comment (vEpiphyte) [inline docs/synapse/userguides/ug013_storm_ref_lift.rst]: broken link

File: docs/synapse/userguides/ug013_storm_ref_lift.rst [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+__ storm.py_
+
+.. _conventions: ../userguides/ug011_storm_basics.rst#syntax-conventions
```
Comment (vEpiphyte) [inline docs/synapse/userguides/ug013_storm_ref_lift.rst]: link to html

File: docs/synapse/userguides/ug014_storm_ref_filter.rst [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+filter()
+--------
+Todo
+
+Filter Comparison Helpers
+-------------------------
+Todo
+
+
+.. _conventions: ../userguides/ug011_storm_basics.rst#syntax-conventions
```
Comment (vEpiphyte) [inline docs/synapse/userguides/ug014_storm_ref_filter.rst]: link to html

File: docs/synapse/userguides/ug014_storm_ref_filter.rst [hunk 13]
```diff
@@ -0,0 +1,20 @@
+Storm Reference - Filter Operators and Comparators
+==================================================
+
+Todo
+
+See the `Storm Syntax Conventions`__ for an explanation of the usage format used below.
+
+Where specific query examples are given, they are commonly provided in pairs using operator syntax followed by the equivalent macro syntax (if available).
+
+filter()
```
Comment (vEpiphyte) [inline docs/synapse/userguides/ug014_storm_ref_filter.rst]: not-usable afaik

File: docs/synapse/userguides/ug013_storm_ref_lift.rst (hunk 10 above)
Comment (vEpiphyte) [inline docs/synapse/userguides/ug013_storm_ref_lift.rst]: link direcly to github master branch of storm.py ?

File: docs/synapse/userguides/ug015_storm_ref_pivot.rst [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+totags()
+--------
+Todo
+
+jointags()
+----------
+Todo
+
+
+.. _storm.py: ../../../synapse/lib/storm.py
```
Comment (vEpiphyte) [inline docs/synapse/userguides/ug015_storm_ref_pivot.rst]: link directly to github master branch of storm.py ?

File: docs/synapse/userguides/ug015_storm_ref_pivot.rst [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+__ storm.py_
+
+.. _conventions: ../userguides/ug011_storm_basics.rst#syntax-conventions
```
Comment (vEpiphyte) [inline docs/synapse/userguides/ug015_storm_ref_pivot.rst]: link to html

File: docs/synapse/userguides/ug016_storm_ref_byhandlers.rst [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+===============================
+Todo
+
+
+See the `Storm Syntax Conventions`__ for an explanation of the usage format used below.
+
+Where specific query examples are given, they are commonly provided in pairs using operator syntax followed by the equivalent macro syntax (if available).
+
+
+.. _conventions: ../userguides/ug011_storm_basics.rst#syntax-conventions
```
Comment (vEpiphyte) [inline docs/synapse/userguides/ug016_storm_ref_byhandlers.rst]: link to html

File: docs/synapse/userguides/ug018_storm_ref_misc.rst [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+---------
+Todo
+
+dset()
+------
+Todo
+
+
+
+.. _conventions: ../userguides/ug011_storm_basics.rst#syntax-conventions
```
Comment (vEpiphyte) [inline docs/synapse/userguides/ug018_storm_ref_misc.rst]: link to html

File: docs/synapse/userguides/ug011_storm_basics.rst [hunk 18]
```diff
@@ -117,11 +233,11 @@ Two ways to ask that question using Storm are:
...
+The first query is problematic because it first asks Storm to return **all** ``inet:ipv4`` nodes within the hypergraph – potentially hundreds of thousands, or even millions of nodes, depending on how densely populated the hypergraph is (mathematically speaking, there are over four billion possible IPv4 addresses). Synapse has to lift **all** of those ``inet:ipv4`` nodes into memory and then select only those nodes with the ``tc.t12`` tag. The query is likely to take an extremely long time to return (at least until query limits are incorporated into Synapse), and therefore represents a "bad" query.
 
 The second query first asks Storm to return **all** nodes tagged with ``tc.t12``. This may still be a large number depending on how much analysis and annotation has been performed related to Threat Group 12. However, the number of nodes tagged ``tc.t12`` will still be much smaller than the number of ``inet:ipv4`` nodes within a hypergraph. As such, the second query is more efficient or performant, and represents a "good" (or at least "better" query).
```
Comment (invisig0th) [inline docs/synapse/userguides/ug011_storm_basics.rst]: "Threat Group 12" might want to change to "Threat Cluster 12" ( continuing the crusade against conflating clustering with attribution ) :D

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst [hunk 19]
```diff
@@ -0,0 +1,299 @@
...
+
+*Separator (sepr) Node:*
+
+``addnode( inet:dns:a , ( woot.com , 1.2.3.4 ) )``
+
+``[ inet:dns:a = ( woot.com , 1.2.3.4 ) ]``
+
+*Composite (comp) Node:*
+
+``addnode( inet:follows , (twitter.com/ernie , twitter.com/bert ) )``
```
Comment (invisig0th) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: lol.  <3

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst [hunk 20]
```diff
@@ -0,0 +1,299 @@
...
+
+**Examples:**
+
+``inet:fqdn = woot.com delnode(force=1)``
+
+**Usage Notes:**
+
+* ``delnode()`` operates on the output of a previous Storm query.
+* Use of the ``force=1`` parameter is meant to require the user to think about what they're doing before executing the ``delnode()`` command (there is no "are you sure?" prompt). Future releases of Synapse will support a permissions structure that will limit the users who are able to execute this operator.
+* ``delnode()`` has the potential to be destructive if executed on an incorrect, badly formed, or mistyped query. Users are strongly encouraged to validate their query by first executing it on its own to confirm it returns the expected nodes before executing ``delnode()``. (Picture accidentally running ``inet:fqdn delnode( force=1 )`` instead of ``inet:fqdn = woot.com delnode( force=1 )``.)
```
Comment (invisig0th) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: Might want to tell them what that Picture means :) in case they can't follow the syntax quite yet.  :D

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst [hunk 21]
```diff
@@ -0,0 +1,299 @@
...
+
+``inet:fqdn = woot.com deltag( baz.faz )``
+
+``inet:fqdn = woot.com [ -#baz.faz ]``
+
+**Usage Notes:**
+
+* ``deltag()`` operates on the output of a previous query.
+* Deleting a leaf tag deletes **only** the leaf tag.
+* Deleting a non-leaf tag deletes that tag and all tags below it in the tag hierarchy.
```
Comment (invisig0th) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: ( and deletes them depth first if you think that helps )

File: docs/synapse/userguides/ug015_storm_ref_pivot.rst (hunk 15 above)
Comment (invisig0th) [inline docs/synapse/userguides/ug015_storm_ref_pivot.rst]: If we were to make all of these links to html, does that break our ability to generate linked PDFs?

File: docs/synapse/userguides/ug011_storm_basics.rst (hunk 2 above)
Comment (therealsilence) [inline docs/synapse/userguides/ug011_storm_basics.rst]: Nah, the point wasn't to "allow" people to use the filter() operator syntax, more a matter of me learning how to structure the docs around what's practical. The docs are kind of set up to continually demonstrate / contrast operator vs macro which just happens to break down here. Not a big deal.

Comment (therealsilence) [inline docs/synapse/userguides/ug011_storm_basics.rst]: Fixed.

File: docs/synapse/userguides/ug011_storm_basics.rst (hunk 3 above)
Comment (therealsilence) [inline docs/synapse/userguides/ug011_storm_basics.rst]: Fixed (removed link as there is no single reference guide to link to).

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst (hunk 4 above)
Comment (therealsilence) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: fixed

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst (hunk 5 above)
Comment (therealsilence) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: fixed

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst (hunk 7 above)
Comment (therealsilence) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: fixed

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst (hunk 8 above)
Comment (therealsilence) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: fixed

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst (hunk 9 above)
Comment (therealsilence) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: clarified / fixed

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst (hunk 20 above)
Comment (therealsilence) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: clarified / fixed

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst (hunk 6 above)
Comment (therealsilence) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: Ooh nice. This looks lame in RST but comes out nicely with sphinx. Thanks for the protip.

File: docs/synapse/userguides/ug013_storm_ref_lift.rst (hunk 10 above)
Comment (therealsilence) [inline docs/synapse/userguides/ug013_storm_ref_lift.rst]: fixed

File: docs/synapse/userguides/ug013_storm_ref_lift.rst (hunk 11 above)
Comment (therealsilence) [inline docs/synapse/userguides/ug013_storm_ref_lift.rst]: fixed

File: docs/synapse/userguides/ug014_storm_ref_filter.rst (hunk 12 above)
Comment (therealsilence) [inline docs/synapse/userguides/ug014_storm_ref_filter.rst]: fixed

File: docs/synapse/userguides/ug014_storm_ref_filter.rst (hunk 13 above)
Comment (therealsilence) [inline docs/synapse/userguides/ug014_storm_ref_filter.rst]: Will still need to describe macro syntax

File: docs/synapse/userguides/ug015_storm_ref_pivot.rst (hunk 14 above)
Comment (therealsilence) [inline docs/synapse/userguides/ug015_storm_ref_pivot.rst]: fixed

File: docs/synapse/userguides/ug015_storm_ref_pivot.rst (hunk 15 above)
Comment (therealsilence) [inline docs/synapse/userguides/ug015_storm_ref_pivot.rst]: ¯\_(ツ)_/¯ ...changing to .html means readthedocs will work correctly. Not sure how that affects PDF.

File: docs/synapse/userguides/ug016_storm_ref_byhandlers.rst (hunk 16 above)
Comment (therealsilence) [inline docs/synapse/userguides/ug016_storm_ref_byhandlers.rst]: fixed

File: docs/synapse/userguides/ug018_storm_ref_misc.rst (hunk 17 above)
Comment (therealsilence) [inline docs/synapse/userguides/ug018_storm_ref_misc.rst]: fixed

File: docs/synapse/userguides/ug011_storm_basics.rst [hunk 22]
```diff
@@ -74,12 +50,7 @@ Storm operators can be divided into broad categories based on their typical use:
 * **Statistical operators** – specialized operators to calculate statistics over a set of nodes.
 * **Miscellaneous operators** – various special purpose operators that do not fit into one of the above categories.
 
-Storm and the Synapse CLI
--------------------------
-
-Recall that when accessing Storm from the Synapse command line, the ``ask`` command indicates that subsequent input represents a Storm query. Storm queries executed from the Synapse CLI must all be preceded by the ``ask`` command:
-
-``cli> ask <query>``
+Most operators (other than those used solely to lift data) require an existing data set on which to operate. This data set is typically the output of a previous Storm query whose results are the nodes you want to modify or otherwise work with.
```
Comment (therealsilence) [inline docs/synapse/userguides/ug011_storm_basics.rst]: fixed

File: docs/synapse/userguides/ug011_storm_basics.rst [hunk 23]
```diff
@@ -93,13 +64,158 @@ Working with Synapse data commonly involves three broad types of operations:
 Whether lifting, filtering, or pivoting across data in a Cortex, you need to be able to clearly specify the data you’re interested in – your selection criteria. In most cases, the criteria you specify will be based on one or more of the following:
 
 * A **property** (primary or secondary) on a node.
-* A **specific value** for a property (``<form>=<value>`` or ``<prop>=<value>``) on a node.
+* A **specific value** for a property (``<form>=<valu>`` or ``<prop>=<pval>``) on a node.
```
Comment (therealsilence) [inline docs/synapse/userguides/ug011_storm_basics.rst]: I'm neutral - have been using <form>=<valu> for primary and <prop>=<pval> for secondary, so did the same here for consistency. Though I don't have strong feelings either way.

File: docs/synapse/userguides/ug011_storm_basics.rst [hunk 24]
```diff
@@ -93,13 +64,158 @@ Working with Synapse data commonly involves three broad types of operations:
...
+
+From an analysis standpoint, this feature means that Storm can parallel an analyst's natural thought process: "show me X data...that's interesting, show me the Y data that relates to X...hm, take only this subset of results from Y and show me any relationship to Z data…" and so on.
+
+From a practical standpoint, it means that **order matters** when constructing a Storm query. A lengthy Storm query is not evaluated as a whole. Instead, Synapse parses each component of the query in order, evaluating each component individually as it goes. The Cortex runtime(s) executing the Storm query keep a list of lifted nodes in memory while performing the requested lifts, pivots, data modification, and so on. The operators used may add or remove nodes from this "working set", or clear the set entirely; as such the in-memory set is continually changing based on the last-used operator. Particularly when first learning Storm, users are encouraged to break down complex queries into their component parts, and to validate the output (results) after the addition of each operator to the overall query.
```
Comment (therealsilence) [inline docs/synapse/userguides/ug011_storm_basics.rst]: Fixed. Thanks to epiphyte :-) for the description.

File: docs/synapse/userguides/ug011_storm_basics.rst [hunk 25]
```diff
@@ -93,13 +64,158 @@ Working with Synapse data commonly involves three broad types of operations:
...
+
+The Synapse documentation provides numerous examples of both abstract Storm syntax (usage statements) and specific Storm queries. The following conventions are used for Storm usage statements:
+
+* Items that must be entered literally on the command line are in **bold.** These items include the command name and literal characters.
+* Items representing variables that must be replaced by a name are in *italics*.
+* **Bold** brackets are literal characters. Parameters enclosed in non-bolded brackets are optional.
+* Parameters **not** enclosed in brackets are required.
+* A vertical bar signifies that you choose only one parameter. For example, ``[ a | b ]`` indicates that you can choose a, b, or nothing.
+* Ellipses ( ``...`` ) signify the parameter can be repeated on the command line.
+
```
Comment (therealsilence) [inline docs/synapse/userguides/ug011_storm_basics.rst]: Thanks to whippit :-)

File: docs/synapse/userguides/ug011_storm_basics.rst [hunk 26]
```diff
@@ -93,13 +64,158 @@ Working with Synapse data commonly involves three broad types of operations:
...
+The most basic Storm query simply lifts (retrieves) a single node (such as the domain ``woot.com``) using the ``lift()`` operator:
+
+``cli> ask lift(inet:fqdn,woot.com)``
+
+The same query can be executed as follows using macro syntax:
+
+``cli> ask inet:fqdn=woot.com``
+
+Note that in macro syntax, the ``lift()`` operator – the most fundamental Storm operator – is eliminated entirely; macro syntax assumes you want to retrieve (lift) nodes unless you specify otherwise. Similarly, instead of entering comma-separated parameters as input to the operator, macro syntax supports the use of the simple ``<form>=<valu>`` pattern to ask about the node in question.
```
Comment (therealsilence) [inline docs/synapse/userguides/ug011_storm_basics.rst]: fixed

File: docs/synapse/userguides/ug011_storm_basics.rst (hunk 18 above)
Comment (therealsilence) [inline docs/synapse/userguides/ug011_storm_basics.rst]: Noted. I use "Threat Group 12" for a number of examples so I'll need to hunt-and-kill in some of the existing docs as well.

File: docs/synapse/userguides/ug011_storm_basics.rst [hunk 27]
```diff
@@ -93,13 +64,158 @@ Working with Synapse data commonly involves three broad types of operations:
...
+
+The full query for this line of analytical reasoning using operator syntax would be::
+
+  cli> ask lift(inet:fqdn,by=tag,tc.t12) pivot(inet:dns:a:fqdn,inet:fqdn) 
+    pivot(inet:ipv4,inet:dns:a:ipv4) -#anon.tor -#anon.vpn 
+    pivot(inet:dns:a:ipv4,inet:ipv4) pivot(inet:fqdn,inet:dns:a:fqdn)
+
+The same query using macro syntax would be::
+
+  cli> ask inet:fqdn*tag=tc.t12 -> inet:dns:a:fqdn inet:dns:a:ipv4 -> inet:ipv4 -#anon.tor -#anon.vpn
```
Comment (therealsilence) [inline docs/synapse/userguides/ug011_storm_basics.rst]: Hmm. Good point. I'm torn; it would be "nice to" show "look how much shorter macro syntax is"! But technically I can also shorten the operator syntax for the pivot() operators, by removing the "from" parameter where the "from" is the form of the node: 

``
ask lift(inet:fqdn,by=tag,tc.t12) pivot(inet:dns:a:fqdn,inet:fqdn)
``

could be:

``
ask lift(inet:fqdn,by=tag,tc.t12) pivot(inet:dns:a:fqdn,)
``

I think if we use the "full" operator syntax (for clarity) we should keep the "full"(ish) macro syntax (since this is a high-level example) and get more into the details for the individual operators and / or tutorial.

File: docs/synapse/userguides/ug015_storm_ref_pivot.rst (hunk 15 above)
Comment (vEpiphyte) [inline docs/synapse/userguides/ug015_storm_ref_pivot.rst]: Haven't looked at PDF generated docs to be honest

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst (hunk 9 above)
Comment (vEpiphyte) [reply docs/synapse/userguides/ug012_storm_ref_datamod.rst]: this is still showing force=1 as required - do we want to change it to be an optional kwarg as per the syntax spec?

Comment (therealsilence) [reply docs/synapse/userguides/ug012_storm_ref_datamod.rst]: fixed - thanks for the catch

File: docs/_static/theme_overrides.css [hunk 28]
```diff
@@ -0,0 +1,13 @@
+/* override table width restrictions */
```
Comment (vertexmc) [inline docs/_static/theme_overrides.css]: All of the css / renaming value->valu, group->cluster looks good. Starting to review the storm docs now.

File: docs/synapse/userguides/ug011_storm_basics.rst [hunk 29]
```diff
@@ -93,35 +64,172 @@ Working with Synapse data commonly involves three broad types of operations:
...
+    pivot(inet:ipv4,inet:dns:a:ipv4) pivot(inet:dns:a:fqdn,inet:fqdn)
+
+The same query using macro syntax would be::
+
+  cli> ask inet:fqdn*tag=tc.t12 -> inet:dns:a:fqdn :ipv4 -> inet:ipv4 -#anon.tor -#anon.vpn
+    -> inet:dns:a:ipv4 :fqdn -> inet:fqdn
+  
+The components of the query are broken down below; note how each new component builds on the previous query to follow the line of analysis and refine results:
+
++-------------------------+----------------------------------------+--------------------------------------------------+
```
Comment (vertexmc) [inline docs/synapse/userguides/ug011_storm_basics.rst]: @therealsilence these tables look awesome! :1st_place_medal:

File: docs/synapse/userguides/ug011_storm_basics.rst [hunk 30]
```diff
@@ -93,35 +64,172 @@ Working with Synapse data commonly involves three broad types of operations:
...
+The second query first asks Storm to return **all** nodes tagged with ``tc.t12``. This may still be a large number depending on how much analysis and annotation has been performed related to Threat Cluster 12. However, the number of nodes tagged ``tc.t12`` will still be much smaller than the number of ``inet:ipv4`` nodes within a hypergraph. As such, the second query is more efficient or performant, and represents a "good" (or at least "better" query).
 
-(**Note:** The previous example is used for simple illustrative purposes. Technically, the "best" way to ask this particular question would be to use what is called a Storm "by" handler (represented by the asterisk ( ``*`` ) to "lift by tag":
+(**Note:** The previous example is used for simple illustrative purposes. Technically, the "best" way to ask this particular question would be to use what is called a Storm "by" handler (represented by the asterisk ( ``*`` )) to "lift by tag":
 
 ``cli> ask inet:ipv4*tag=tc.t12``
```
Comment (vertexmc) [inline docs/synapse/userguides/ug011_storm_basics.rst]: This file looked good to me

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst [hunk 31]
```diff
@@ -0,0 +1,332 @@
...
+* Add or modify properties (``setprop()``).
+* Delete properties (once ``delprop()`` is implemented).
+* Add tags (``addtag()``).
+* Delete tags (``deltag()``).
+
+This means that all of the above directives can be specified within a single set of macro syntax brackets, in any combination and in any order.
+
+However, it is important to keep in mind that **the brackets are NOT a boundary that segregates nodes.** The brackets simply indicate the start and end of data modification shorthand. They do **NOT** separate "nodes these modifications should apply to" from "nodes they should not apply to". The Storm `operator chaining`__ with left-to-right processing order still applies. Any modification request that operates on previous Storm output will operate on the output of everything “leftwards” of the modifier, regardless of whether that content is within or outside of the macro syntax brackets. For example:
+
+``inet:ipv4 = 12.34.56.78 inet:fqdn = woot.com [ inet:ipv4 = 1.2.3.4 :created = "2016-12-18 00:35" inet:fqdn = woowoo.com #my.tag ]``
```
Comment (vertexmc) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: This line goes off the page for me. Did we find a way to fix this before?

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst [hunk 32]
```diff
@@ -0,0 +1,332 @@
...
+.. _ingest: ../userguides/ug050_ing_intro.html
+__ ingest_
+
+.. _xref: ../userguides/ug007_dm_nodetypes.html#cross-reference-xref-nodes
+__ xref_
+
+.. _timestamps: ../userguides/ug008_dm_tagconcepts.html#tag-timestamps
+
+.. _chaining: ../userguides/ug011_storm_basics.html#operator-chaining
+__ chaining_
```
Comment (vertexmc) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: This page looks good to me, other than the one note I left.

File: docs/synapse/userguides/ug013_storm_ref_lift.rst [hunk 33]
```diff
@@ -0,0 +1,27 @@
...
+Todo
+
+alltag()
+--------
+Todo
+
+.. _storm.py: https://github.com/vertexproject/synapse/blob/master/synapse/lib/storm.py
+
+.. _conventions: ../userguides/ug011_storm_basics.html#syntax-conventions
+__ conventions_
```
Comment (vertexmc) [inline docs/synapse/userguides/ug013_storm_ref_lift.rst]: This page looks good to me as long as the PRs will be done in another PR (and weren't intended to be done in this one)

File: docs/synapse/userguides/ug014_storm_ref_filter.rst (hunk 13 above)
Comment (vertexmc) [reply docs/synapse/userguides/ug014_storm_ref_filter.rst]: I would note that `filter` is not actually an operator and that users need to use `<plus or minus><prop><optional comparator and valu>` or `<plus or minus><#tagname>`.

File: docs/synapse/userguides/ug015_storm_ref_pivot.rst [hunk 34]
```diff
@@ -0,0 +1,40 @@
...
+
+jointags()
+----------
+Todo
+
+
+.. _storm.py: https://github.com/vertexproject/synapse/blob/master/synapse/lib/storm.py
+
+.. _conventions: ../userguides/ug011_storm_basics.html#syntax-conventions
+__ conventions_
```
Comment (vertexmc) [inline docs/synapse/userguides/ug015_storm_ref_pivot.rst]: This page looks good to me as long as the PRs will be done in another PR (and weren't intended to be done in this one)

File: docs/synapse/userguides/ug016_storm_ref_byhandlers.rst [hunk 35]
```diff
@@ -0,0 +1,12 @@
...
+Todo
+
+
+See the `Storm Syntax Conventions`__ for an explanation of the usage format used below.
+
+Where specific query examples are given, they are commonly provided in pairs using operator syntax followed by the equivalent macro syntax (if available).
+
+
+.. _conventions: ../userguides/ug011_storm_basics.html#syntax-conventions
+__ conventions_
```
Comment (vertexmc) [inline docs/synapse/userguides/ug016_storm_ref_byhandlers.rst]: This page looks good to me as long as the PRs will be done in another PR (and weren't intended to be done in this one)

File: docs/synapse/userguides/ug017_storm_ref_stats.rst [hunk 36]
```diff
@@ -0,0 +1,16 @@
...
+
+Where specific query examples are given, they are commonly provided in pairs using operator syntax followed by the equivalent macro syntax (if available).
+
+stat()
+------
+Todo
+
+
+.. _conventions: ../userguides/ug011_storm_basics.html#syntax-conventions
+__ conventions_
```
Comment (vertexmc) [inline docs/synapse/userguides/ug017_storm_ref_stats.rst]: This page looks good to me as long as the PRs will be done in another PR (and weren't intended to be done in this one)

File: docs/synapse/userguides/ug018_storm_ref_misc.rst [hunk 37]
```diff
@@ -0,0 +1,45 @@
...
+Todo
+
+dset()
+------
+Todo
+
+
+
+.. _conventions: ../userguides/ug011_storm_basics.html#syntax-conventions
+__ conventions_
```
Comment (vertexmc) [inline docs/synapse/userguides/ug018_storm_ref_misc.rst]: This page looks good to me as long as the PRs will be done in another PR (and weren't intended to be done in this one)

File: docs/synapse/userguides/ug014_storm_ref_filter.rst [hunk 38]
```diff
@@ -0,0 +1,20 @@
...
+--------
+Todo
+
+Filter Comparison Helpers
+-------------------------
+Todo
+
+
+.. _conventions: ../userguides/ug011_storm_basics.html#syntax-conventions
+__ conventions_
```
Comment (vertexmc) [inline docs/synapse/userguides/ug014_storm_ref_filter.rst]: This page looks good to me, other than the one note I left.

File: docs/synapse/userguides/ug011_storm_basics.rst (hunk 29 above)
Comment (therealsilence) [reply docs/synapse/userguides/ug011_storm_basics.rst]: They were a pain. @vEpiphyte helped a lot. :-)

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst [hunk 39]
```diff
@@ -42,38 +42,41 @@ Adds the specified node(s) to a Cortex.
 **Examples:**
 
 *Simple Node:*
-
-``addnode( inet:fqdn , woot.com )``
-
-``[ inet:fqdn = woot.com ]``
+::
```
Comment (vertexmc) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: I think it looks a lot better after this change

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst [hunk 40]
```diff
@@ -1,3 +1,5 @@
+.. highlight:: none
```
Comment (vertexmc) [inline docs/synapse/userguides/ug012_storm_ref_datamod.rst]: Woot! easier to read

File: docs/synapse/userguides/ug012_storm_ref_datamod.rst (hunk 31 above)
Comment (therealsilence) [reply docs/synapse/userguides/ug012_storm_ref_datamod.rst]: Fixed per internal discussion.
